romtools/linalg/linalg.py

```diff
@@ -192,6 +192,142 @@ def _basic_max_via_python(a: np.ndarray, axis=None, comm=None):
     return local_max
 
 
+# ----------------------------------------------------
+def _basic_argmax_via_python(a: np.ndarray, comm=None):
+    '''
+    Return the index of an array's maximum value. If the array is distributed, also returns the
+    value itself and the MPI rank on which it occurs.
+
+    Parameters:
+        a (np.ndarray): input data
+        comm (MPI_Comm): MPI communicator (default: None)
+
+    Returns:
+        if comm == None, returns the index of the maximum value (identical to np.argmax)
+        if comm != None, returns a tuple containing (value, index, rank):
+            value: the global maximum
+            index: the local index of the global maximum
+            rank:  the rank on which the global maximum resides
+
+    Preconditions:
+      - a is at most a rank-3 tensor
+      - if a is a distributed 2-D array, it must be distributed along axis=0,
+        and every rank must have the same a.shape[1]
+      - if a is a distributed 3-D tensor, it must be distributed along axis=1,
+        and every rank must have the same a.shape[0] and a.shape[2]
+
+    Postconditions:
+      - a and comm are not modified
+
+    Example 1:
+    **********
+
+       rank 0  2.2
+               3.3
+      =======================
+       rank 1  40.
+               51.
+               -24.
+               45.
+      =======================
+       rank 2  -4.
+
+    Suppose that we do:
+
+        res = la.argmax(a, comm)
+
+    then ALL ranks will contain res = (51., 1, 1).
+    (The global maximum is 51., and it occurs at index 1 of the local array on Rank 1.)
+
+    Example 2:
+    **********
+
+       rank 0  2.2  1.3  4.
+               3.3  5.0  33.
+      =======================
+       rank 1  40.  -2.  -4.
+               51.   4.   6.
+               -24.  8.   9.
+               45.  -3.  -4.
+      =======================
+       rank 2  -4.  8.   9.
+
+    Suppose that we do:
+
+       res = la.argmax(a, comm)
+
+    then ALL ranks will contain res = (51,. 3, 1)
+    (The maximum value is 51., and it occurs at index 3 of the flattened local array on Rank 1.)
```

Comment:
does flattening depend on the memory layout? or is is alawsy row-wise flattening?

Reply:
The default for np.ravel is row-major (see docs [here](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html)), but we can specify the memory layout if desired.

The `"A"` option seems to depend on the memory layout of the array:

>  ‘A’ means to read the elements in Fortran-like index order if a is Fortran contiguous in memory, C-like order otherwise.